src/github_repo_request_local.py

```diff
@@ -83,45 +164,154 @@ def list_test_files(directory, excluded_extensions):
     else:
         logger.error(f"CSV file not found at {csv_file_path}.")
 
+# Filter out rows where the URL doesn't have a repository name (83 rows)
+if "repourl" in df.columns:
+    # Identify rows with incomplete URLs
+    incomplete_urls = df[
+        df["repourl"].apply(lambda x: len(x.rstrip("/").split("/")) < 5)
+    ]
 
-for index, row in df.iterrows():
-    if row["testfilecountlocal"] != -1:
-        continue  # Skip processed repositories
+    # Log the incomplete URLs
+    if not incomplete_urls.empty:
+        logger.info(
+            f"{len(incomplete_urls)} incomplete GitHub URLs found and "
+            f"will be excluded:"
+        )
+        for url in incomplete_urls["repourl"]:
+            logger.info(f"Excluding the repourl : {url}")
 
-    repo_url = row["repourl"]
-    repo_name = Path(repo_url.split("/")[-1]).stem
-    clone_dir = clone_dir_base / repo_name
+    df = df[df["repourl"].apply(lambda x: len(x.rstrip("/").split("/")) >= 5)]
 
-    if not clone_dir.exists():
-        try:
-            logger.info(f"Trying to clone {repo_url} into {clone_dir}")
-            subprocess.run(
-                ["git", "clone", repo_url, str(clone_dir)],
-                check=True,
-                capture_output=True,
-            )
-            logger.info(f"Successfully cloned the repo: {repo_name}")
-        except subprocess.CalledProcessError as e:
-            logger.error(
-                f"Failed to clone the repo: {repo_name}." f" " f"Exception: {e}"
-            )
-            df.at[index, "testfilecountlocal"] = 0
+
+if "last_commit_hash" not in df.columns:
+    df["last_commit_hash"] = None  # Initialize the column with None
+
+# replace http with https
+df["repourl"] = df["repourl"].str.replace(r"^http\b", "https", regex=True)
+
+# Some of the URLs end with "/". I need to remove them.
+df["repourl"] = df["repourl"].str.rstrip("/")
+
+# Clean and filter URLs
+df["repourl"] = df["repourl"].apply(get_base_repo_url)
+df = df[df["repourl"] != ""]  # Remove any rows with invalid URLs
+
+# Number of repositories to process before saving to CSV
+BATCH_SIZE = 10
+
+# Track the number of processed repositories in the current batch
+processed_count = 0
+
+# Get the path to save the output file
+output_file_path = repo_root / "data" / "repo_testfiles_output.txt"
+
+# Redirect stdout to a file for outputting repo URLs and test file names
+original_stdout = sys.stdout
```

Comment:
Generally I don't expect programs to redirect stdout, who/what ever runs the script makes that choice. The might want to 'see' stdout on the console, redirect it to a file, or do both (e.g. using the `tee` command line utility). In the code we can trivially create and open a file and then write to it, etc. So if we want to always write to a file then we can do so without redirecting stdout. 

How about taking the second approach here - of writing to a file (and remove the redirection of stdout in the code)?